From 50b884c43fc5970b4d70d112651fe42319465bb3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Tue, 21 Oct 2025 14:59:36 +0200
Subject: [PATCH 21/21] Disable {position,resize}Automatic on first expose

Disabling automatic positioning/resize after platform
window creation assumed that the platform window could
synchronously determine the automatic position/size,
either via QPlatformWindow::initialGeometry(), or via
platform APIs.

But for platforms like XCB, the automatic position is
determined by the window manager when the window is
first mapped. With the existing logic, we would end up
overriding the position when QWidget would set size
hints or resize the platform window during setVisible.

We now defer the reset of {position,resize}Automatic
until the expose event. Ideally we should do it only
when the window is first shown, but we don't have a
QWSI for tracking window visible-state.

Amends 59d77a0162d92a9993b7bb4c593e88cbb8f8d9c8.

Fixes: QTBUG-141099
Pick-to: 6.10
Change-Id: I53a0205c7fbddf922b030b3f4bd9b6ce5c4915b9
Reviewed-by: Liang Qi <liang.qi@qt.io>
---
 src/gui/kernel/qguiapplication.cpp | 9 +++++++++
 src/gui/kernel/qwindow.cpp         | 6 ------
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/src/gui/kernel/qguiapplication.cpp b/src/gui/kernel/qguiapplication.cpp
index 5fd48a55344..0b4cfc4249f 100644
--- a/src/gui/kernel/qguiapplication.cpp
+++ b/src/gui/kernel/qguiapplication.cpp
@@ -3417,6 +3417,15 @@ void QGuiApplicationPrivate::processExposeEvent(QWindowSystemInterfacePrivate::E
         return;
     QWindowPrivate *p = qt_window_private(window);
 
+    if (e->isExposed) {
+        // If the window has been automatically positioned or resized by the
+        // window manager, we now assume those have taken effect, even for
+        // asynchronous window managers. From this point on we want the window
+        // to keep its geometry, even when recreated.
+        p->positionAutomatic = false;
+        p->resizeAutomatic = false;
+    }
+
     if (!p->receivedExpose) {
         if (p->resizeEventPending) {
             // as a convenience for plugins, send a resize event before the first expose event if they haven't done so
diff --git a/src/gui/kernel/qwindow.cpp b/src/gui/kernel/qwindow.cpp
index d3d98eee6fe..d9d517d5790 100644
--- a/src/gui/kernel/qwindow.cpp
+++ b/src/gui/kernel/qwindow.cpp
@@ -586,12 +586,6 @@ void QWindowPrivate::create(bool recursive)
 
     platformWindow->initialize();
 
-    // Now that the window is created and initialized the platform has had
-    // a chance to position and size it automatically. From this point on
-    // we want the window to keep its geometry, even when recreated.
-    positionAutomatic = false;
-    resizeAutomatic = false;
-
     QObjectList childObjects = q->children();
     for (int i = 0; i < childObjects.size(); i ++) {
         QObject *object = childObjects.at(i);
-- 
2.51.1

